cupy/array_api/_data_type_functions.py

```diff
@@ -96,7 +94,7 @@ def finfo(type: Union[Dtype, Array], /) -> finfo_object:
 
     See its docstring for more information.
     """
-    fi = np.finfo(type)
+    fi = np.finfo(type)  # type: ignore
```

Comment:
The errors for `finfo` and `iinfo` are more legit though. The issue stems from the `Array.dtype` property being annotated as [returning a union](https://github.com/cupy/cupy/blob/ec6d2c9a98440b19a133868576fb299972c21f6c/cupy/array_api/_typing.py#L56-L67), which _always_ includes both floating- and integer-types while the `iinfo`/`finfo` constructors only accept one category or the other. 
This is a demonstration why it's usually a bad idea to return these so-called "unsafe" unions, and I would recommend to simply switch the dtype parameter over to `Any`:

``` diff
diff --git a/cupy/array_api/_array_object.py b/cupy/array_api/_array_object.py
index c2ae0545d..d713e927a 100644
--- a/cupy/array_api/_array_object.py
+++ b/cupy/array_api/_array_object.py
@@ -1047,7 +1047,7 @@ class Array:
             return Array._new(arr)

     @property
-    def dtype(self) -> Dtype:
+    def dtype(self) -> np.dtype[Any]:
         """
         Array API compatible wrapper for :py:meth:`np.ndarray.dtype <numpy.ndarray.dtype>`.
```